tensorflow/core/kernels/sparse_tensor_dense_matmul_op_gpu.cu.cc

```diff
@@ -71,60 +101,111 @@ __global__ void SparseTensorDenseMatMulKernel(
     // b_value == (ADJ_B) ? conj(b[j, k]) : b[k, j]
     const T b_input = ldg(b + ((ADJ_B) ? j * b_cols + k : k * b_cols + j));
     const T b_value = ADJ_B ? Eigen::numext::conj(b_input) : b_input;
-    GpuAtomicAdd(out_location,
-                 static_cast<Tsum>(a_value) * static_cast<Tsum>(b_value));
+
+    GpuRecastAtomicAdd<Tsum, T>(out_location, a_value * b_value);
   }
 }
 
 namespace functor {
 
+namespace {
+
+bool RequireDeterminism() {
+  static bool require_determinism = [] {
+    bool deterministic_ops = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_DETERMINISTIC_OPS",
+                                               /*default_val=*/false,
+                                               &deterministic_ops));
+    return deterministic_ops;
+  }();
+  return require_determinism;
+}
+
+} // namespace
+
 template <typename T, typename Tindices, bool ADJ_A, bool ADJ_B>
 struct SparseTensorDenseMatMulFunctor<GPUDevice, T, Tindices, ADJ_A, ADJ_B> {
+
+  template <typename Tsum>
   static EIGEN_ALWAYS_INLINE Status
-  Compute(OpKernelContext* ctx, typename TTypes<T>::Matrix out,
-          typename TTypes<Tindices>::ConstMatrix a_indices,
-          typename TTypes<T>::ConstVec a_values,
-          typename TTypes<T>::ConstMatrix b) {
+  GpuSparseTensorDenseMatMulWithRecast(
```

Comment:
Nice move to put this inside the functor. That provides simplification opportunities because it limits the scope and context and keeps the solution well encapsulated (including not polluting the namespace above).

However, the function does not need this name now because it's within the context of `SparseTensorDenseMatMulFunctor`.

What does it do? It performs a subset of the functionality of `SparseTensorDenseMatMulFunctor::Compute`. It performs the functionality of `Compute` after `Tsum` has been defined. `Tsum` is a potentially higher precision type used for the accumulation.

I don't think "recast" is actually a good summary for what is happening here.

**I think that `ComputeWithPossiblyHigherPrecisionAccumulator` would be an accurate, descriptive, complete, and unambiguous name for this function.**

Note that I used the term "higher-precision accumulator" rather than the shorter "super-accumulator". I think that "higher-precision accumulator" provides a clearer description of what's happening than "super-accumulator," which is an esoteric computer architecture term that people may not have heard before.

**Also, this function should be `private` since it should never be called from outside the class.**

Reply:
Note to self: both parts of this have been addressed.